src/NBitcoin/ChainedHeader.cs

```diff
@@ -4,6 +4,60 @@
 
 namespace NBitcoin
 {
+    /// <summary>
+    /// Represents the state of the availability of a block.
+    /// </summary>
+    public enum BlockDataAvailabilityState
+    {
+        /// <summary>
+        /// A <see cref="BlockHeader"/> is found and is a valid header.
+        /// </summary>
+        HeaderOnly,
+
+        /// <summary>
+        /// We are interested in downloading the <see cref="Block"/> that is being represented by the current <see cref="BlockHeader"/>. 
+        /// This happens when a header is a part of the chain that can potentially replace our consensus tip 
+        /// because its tip's total chain work is greater comparing to what we currently have.
+        /// </summary>
+        BlockRequired,
+
+        /// <summary>
+        /// The <see cref="Block"/> is downloaded, if the Block property is null the block needs to be asked from store.
+        /// </summary>
+        BlockAvailable
+    }
+
+    /// <summary>
+    /// Represents the validation level of a block.
+    /// </summary>
+    public enum ValidationState
+    {
+        /// <summary>
+        /// No validation was performed.
+        /// </summary>
+        Unknown,
+
+        /// <summary>
+        /// We have a valid block header.
+        /// </summary>
+        HeaderValidated,
+
+        /// <summary>
+        /// No validation required because the block is assumed to be valid.
+        /// </summary>
+        AssumedValid,
+
+        /// <summary>
+        /// Validated using all rules that don't require db state changes.
```

Comment:
However, for a block that was represented using headers with assume valid state, this state does not mean that it was validated using all the rules, but that the rules were skipped because of the assumption of validity.

Reply:
It wont necessarily be inputs though, I think its more about the state of the chain, partial validation will not change any state (for example outputs will not be marked as spent) this will only do validation, in case of smart contracts its not clear yet.

Reply:
connecting inputs (newly called UpdateCoins) means that we check that inputs are UTXOs (this can only be done after we rewind the chain to the fork point) and we spend them while creating new UTXOs from new outputs, that is the last step. Bitcoin core does a few extra checks in this phase - like sigop count and some BIP30/34 thingy, but more or less, it is just pure input connecting.

Reply:
Though there is one more thing I need to check. So wait for it.